### PR TITLE
Added THREE.Timer.

### PR DESCRIPTION
**Pull request by Mugen87**
Related #17679

This is a new alternative to `THREE.Clock` with a different API design and behavior. It should make it easier to implement `three.js` apps FPS independent which becomes more and more important with faster refresh rates.

Background: Some time ago, I've realized on a 144 Hz gaming laptop that (at least in Chrome and Firefox) WebGL apps run with 144 FPS. Some `three.js` examples now look weird because they assume fix update steps (around 60 FPS). Since certain laptops already offer 240 Hz screens, I think we should reconsider how a time class works. The main differences of `THREE.Timer` compared to `THREE.Clock` are:

- `THREE.Timer` has an update method that updates its internal state.
- That makes it possible to call `.getDeltaTime()` and `.getElapsedTime()` multiple times per simulation step without getting different values.
-  The class uses the Page Visibility API to avoid large time delta values when the app is inactive (e.g. tab switched or browser hidden).
- It's possible to configure a fixed time delta and a time scale value (similar to Unity's `Time` interface).
- No public properties, no ctor parameter. The public API only consists of methods.

`THREE.Timer` is located in `examples/jsm/misc/` for now. I personally think `THREE.Timer` (or a similar class) should replace `THREE.Clock` at some point since the class has some conceptual design flaws. The example [webgl_morphtargets_sphere](https://raw.githack.com/Mugen87/three.js/dev38/examples/webgl_morphtargets_sphere.html) already uses `THREE.Timer` in order to replace the existing FPS workaround.

The method names are quite descriptive now. If e.g. `getDelta()` is preferred over `getDeltaTime()`, I'm happy to shorten the names.

### CODE REVIEW

**Comment by fernandojsg** [comment]
I like it! Just a question, what is the expected use of `now()` if fixedTime is enabled? maybe is safer to return elapsedTime in that case? as it could be inconsistent to use `now()` together with the elapsedTime provided by the fixed steps.

**Comment by Mugen87** [comment]
> Just a question, what is the expected use of now() if fixedTime is enabled?

It's probably better to make `now()` private in order to avoid this confusion. The method should just return a time stamp which is used to compute the delta time when no fixed deltas are used. 

`Timer.now()` should not be affected by `timeScale`.  If `elapsedTime` is used in `now()`, that would be the case.

**Comment by fernandojsg** [comment]
So in that case I agree on making `now()` private

File: examples/js/misc/Timer.js [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+	function Timer() {
+
+		this._previousTime = 0;
+		this._currentTime = 0;
+
+		this._deltaTime = 0;
+		this._elapsedTime = 0;
+
+		this._timeScale = 1;
```
Comment (looeee) [inline examples/js/misc/Timer.js]: Nit: timescale is a word, there's no need to capitalize Scale

**Comment by looeee** [comment]
>  If getDelta() is preferred over getDeltaTime(), I'm happy to shorten the names.

I think `getDelta()` is fine - in the context of a timer, it's very clear that we are talking about time deltas.

File: examples/js/misc/Timer.js (hunk 1 above)
Comment (Mugen87) [reply examples/js/misc/Timer.js]: 👍

**Comment by looeee** [comment]
In what situations would you want to use `fixedDeltaTime`?

**Comment by Mugen87** [comment]
It's actually useful for testing and debugging if a timer that can produce deterministic values.

**Comment by Mugen87** [comment]
> I think getDelta() is fine - in the context of a timer, it's very clear that we are talking about time deltas.

Okay, I've shorten the method names and variables.

**Comment by donutcoffee** [comment]
Nice.

Is there a reason you've chosen to add a separate "update()" call?

I.e. why not update the timer when "getDelta()" is called like the current Clock?

**Comment by Mugen87** [comment]
> I.e. why not update the timer when "getDelta()" is called like the current Clock?

Because many users complained about the fact that multiple calls of `getDelta()` within a single frame produces wrong results. `getElapsedTime()` has similar issues, see #5696.

To avoid this design problems, the state of the timer is now explicitly updated by calling `update()`. Yes, it is an additional method call but it avoids confusing behaviors of the current `Clock` class.

**Comment by LeviPesin** [comment]
Should also the Timer class have the serialization/deserialization methods? They would be nice to have.

**Comment by Mugen87** [comment]
Sserialization/deserialization is only added to core classes which are part of the scene graph. `Clock` and `Timer` and not added to the scene.

**Comment by LeviPesin** [comment]
@Mugen87 When this PR would be merged?

**Comment by Mugen87** [comment]
@LeviPesin TBH, I don't think it's a good idea to post at PRs asking for when are they getting merged or if merge conflicts can be resolved. PRs are paused/stalled for various reasons. Sometimes the general approach isn't right, sometimes collaborators are not happy with certain parts or have a different solutions in mind. However, not every collaborator has the time or opportunity to explain their concerns in detail because other tasks are more important.

Looking at the past, many stalled PR were merged when the team was confident enough with the related design decision.

**Comment by drcmda** [comment]
we started to add it to three-stdlib, i noticed that the class has constructor side effects, for reference: https://github.com/mrdoob/three.js/issues/20575

i've added a connect method that pairs with dispose https://github.com/pmndrs/three-stdlib/blob/main/src/misc/Timer.ts that would also make the object pausable/re-usable.

**Comment by vanruesc** [comment]
I've also been using a modified [Timer](https://github.com/pmndrs/postprocessing/blob/92b0cd7b6a69186e5ea7be3408ede9d87de50f8a/src/utils/Timer.ts) class in [@pmndrs/postprocessing](https://github.com/pmndrs/postprocessing) which could serve as feedback for the API design.

This version uses getter properties instead of methods and has an `autoReset` setting that controls usage of the page visibility API. The `update` method also has an optional [timestamp](https://github.com/pmndrs/postprocessing/blob/92b0cd7b6a69186e5ea7be3408ede9d87de50f8a/src/utils/Timer.ts#L156) parameter which can be obtained from the requestAnimationFrame callback to avoid unnecessary calls to `performance.now`:

```ts
requestAnimationFrame(function render(timestamp: number): void {

	requestAnimationFrame(render);
	timer.update(timestamp);

});
```

**Comment by Mugen87** [comment]
> The update method also has an optional timestamp parameter which can be obtained from the requestAnimationFrame callback to avoid unnecessary calls to performance.now:

That is a good idea! I'll add it to the PR.

**Comment by manake** [comment]
I think there may be a bug in `_now()` function.

```
let timer = new Timer();

// Let it run to for example 10 seconds.

timer = null;

// Wait 10 seconds.

timer = new Timer();

timer.getElapsed(); // returns 0 (as expected)
timer.update();
timer.getElapsed(); // returns 20, so there is a sudden gap of 20 seconds (unexpected)
```

Shouldn't it count from 0 after the second initialization?

**Comment by Mugen87** [comment]
For reproduction: https://jsfiddle.net/3kejsft2/

Indeed! This happens because `performance.now()` returns a value which is relative to the application start. I guess I have designed `Timer` without the assumption somebody wants to creation multiple instances of it.

Can you please verify if this version of `Timer` fixes your issue? https://jsfiddle.net/3kejsft2/1/

The idea is to save the value of `performance.now()` when the timer is created and then just subtract it.

**Comment by Mugen87** [comment]
In the meanwhile, `performance.now()` is so well supported I guess we can remove the fallback to `Date`.

**Comment by manake** [comment]
Thank you!

It seems to work now.

(There's also a minor inconsistency that `if ( document.hidden === false ) this.reset();` is lacking curly braces but all other if statements have them (one-liners and multiple-liners).)

**Comment by ycw** [comment]
would you mind elaborate that how the Timer solves this "design flaw" of Clock https://github.com/mrdoob/three.js/pull/26806#issuecomment-1727706775? any help is appreciated.

**Comment by Mugen87** [comment]
Updating the timer per frame before querying data is the intended usage.

**Comment by Mugen87** [comment]
I've seen projects in the past copying this code or use an adapted version of it so I think it's better to provide an alternative to `THREE.Clock` in `three.js`.

**Comment by vanruesc** [comment]
> > The update method also has an optional timestamp parameter which can be obtained from the requestAnimationFrame callback to avoid unnecessary calls to performance.now:
> 
> That is a good idea! I'll add it to the PR.

Just curious: why was this removed in d443665? :thinking:

**Comment by WestLangley** [comment]
>I personally think THREE.Timer (or a similar class) should replace THREE.Clock at some point 

How about adding it to core now, with documentation? Then, deprecate `THREE.Clock` eventually.

If you don't want to do that now, how about adding the above comprehensive description to the header of `Timer.js`?

**Comment by Mugen87** [comment]
> Just curious: why was this removed in https://github.com/mrdoob/three.js/commit/d443665a59479f728f7de41ac58a9dd4254a4c91? 🤔

I was unsure about this feature since the parameter has no effect when a fixed time delta is used. 

> If you don't want to do that now, how about adding the above comprehensive description to the header of Timer.js?

That sounds good to me!

**Comment by Mugen87** [comment]
> > If you don't want to do that now, how about adding the above comprehensive description to the header of Timer.js?

> That sounds good to me!

Done! 46b6bb0ab34427b36bc25c4747db2bcdb88905bf

For `r161`, I'll try to add a documentation page.

**Comment by mrdoob** [comment]
Sorry for the huge delay dealing with this... 🙏

File: examples/jsm/misc/Timer.js [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+		this._currentTime = 0;
+		this._startTime = now();
+
+		this._delta = 0;
+		this._elapsed = 0;
+
+		this._timescale = 1;
+
+		this._useFixedDelta = false;
+		this._fixedDelta = 16.67; // ms, corresponds to approx. 60 FPS
```
Comment (mrdoob) [inline examples/jsm/misc/Timer.js:15]: Hmm, my current phone runs at 120fps... 🤔

File: examples/webgl_morphtargets_sphere.html [hunk 3]
```diff
@@ -52,7 +53,7 @@
 
 				scene = new THREE.Scene();
 
-				clock = new THREE.Clock();
+				timer = new Timer();
```
Comment (mrdoob) [inline examples/webgl_morphtargets_sphere.html:56]: Extra spaces here.

File: examples/webgl_morphtargets_sphere.html [hunk 4]
```diff
@@ -117,30 +116,17 @@
...
 				requestAnimationFrame( animate );
+				timer.update();
 				render();
 
 			}
 
 			function render() {
 
-				const delta = clock.getDelta();
+				const delta = timer.getDelta();
```
Comment (mrdoob) [inline examples/webgl_morphtargets_sphere.html:129]: Extra spaces here too.

File: examples/jsm/misc/Timer.js [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+	}
+
+	enableFixedDelta() {
+
+		this._useFixedDelta = true;
+
+		return this;
+
+	}
+
```
Comment (mrdoob) [inline examples/jsm/misc/Timer.js:58]: I think we can merge these methods into one:

```js
	setFixedDelta( boolean ) {

		this._useFixedDelta = boolean;

		return this;

	}
```

**Comment by mrdoob** [comment]
Hmmm... I'm not sure about the API for fixedTime.
I'll open a new PR for it... 🤔

**Pull request merged**